signac/core/collection_api.py

```diff
@@ -0,0 +1,491 @@
+from collections.abc import Sequence
+from collections.abc import MutableSequence
+from abc import abstractmethod
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+try:
+    from collections.abc import Collection
+except ImportError:
+    # Collection does not exist in Python 3.5, only Python 3.6 or newer.
+
+    from collections.abc import Sized, Iterable, Container
+
+    def _check_methods(C, *methods):
+        mro = C.__mro__
+        for method in methods:
+            for B in mro:
+                if method in B.__dict__:
+                    if B.__dict__[method] is None:
+                        return NotImplemented
+                    break
+            else:
+                return NotImplemented
+        return True
+
+    class Collection(Sized, Iterable, Container):  # type: ignore
+        @classmethod
+        def __subclasshook__(cls, C):
+            if cls is Collection:
+                return _check_methods(C,  "__len__", "__iter__", "__contains__")
+            return NotImplemented
+
+
+class SyncedCollection(Collection):
+
+    def __init__(self, parent=None):
+        self._data = None
+        self._suspend_sync_ = 0
+        self._parent = parent
+
+    # TODO add back-end
+    @classmethod
+    def from_base(self, data, filename=None, parent=None):
+        if isinstance(data, Mapping):
+            return JSONDict(filename=filename, data=data, parent=parent)
+        elif isinstance(data, Sequence) and not isinstance(data, str):
+            return JSONList(filename=filename, data=data, parent=parent)
+        elif NUMPY:
+            if isinstance(data, numpy.number):
+                return data.item()
+            elif isinstance(data, numpy.ndarray):
+                return JSONList(filename=filename, data=data.tolist(), parent=parent)
+        return data
+
+    @abstractmethod
+    def to_base(self):
+        pass
+
+    @contextmanager
+    def _suspend_sync(self):
+        self._suspend_sync_ += 1
+        yield
+        self._suspend_sync_ -= 1
+
+    @abstractmethod
+    def _load(self):
+        pass
+
+    @abstractmethod
+    def _sync(self):
+        pass
+
+    @contextmanager
+    def _safe_sync(self):
+        backup = self._data
+        try:
+            yield
+        except BaseException:
+            self._data = backup
+            raise
+
+    def sync(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                self._sync()
+            else:
+                self._parent.sync()
+
+    def load(self):
+        if self._suspend_sync_ <= 0:
+            if self._parent is None:
+                data = self._load()
+                with self._suspend_sync():
+                    self.reset(data)
+            else:
+                self._parent.load()
+
+
+class _SyncedDict(SyncedCollection, MutableMapping):
+
+    _PROTECTED_KEYS = ('_data', '_suspend_sync_', '_load', '_sync', '_parent')
```

Comment:
Yes, this is valid and I think we might benefit from constructing this list programatically rather than keeping it manually up to date. The purpose is to avoid conflicts between setting, getting and deleting elements of the dictionary and attributes that are actually methods of a dictionary.